File: Soal2/Nomor2_063_078_090.java
/*
Nama Program  : 
Nama Kelompok : Tim PeBO
Nama Anggota  : Razan Ibrahim Nabil, Muhammah Irsyad Azzarul Haq, Djeremy Rieldy Marchiano Panjaitan
NPM  Anggota  : 140810250090, 250078, 250063
Tanggal Buat  : 24/09/2026
Deskripsi     : 
*/

import java.time.Duration;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.Scanner;

class Pegawai{
    private static final DateTimeFormatter FORMAT_WAKTU = DateTimeFormatter.ofPattern("HH:mm");

    private String nip;
    private String nama;
    private int golongan;
    private LocalTime waktuDatang;
    private LocalTime waktuPulang;

    public Pegawai(){
        this.nip = "";
        this.nama = "";
        this.golongan = 0;
        this.waktuDatang = null;
        this.waktuPulang = null;
    }

    public Pegawai(String nip, String nama, int golongan){
        this.nip = nip;
        this.nama = nama;
        this.golongan = golongan;
        this.waktuDatang = null;
        this.waktuPulang = null;
    }

    public void setPegawai(String nip, String nama, int golongan){
        this.nip = nip;
        this.nama = nama;
        this.golongan = golongan;
    }

    public void setNip(String nip) {
        this.nip = nip;
    }

    public void setNama(String nama) {
        this.nama = nama;
    }

    public void setGolongan(int golongan) {
        this.golongan = golongan;
    }

    public void setWaktuDatang(String waktu) {
        this.waktuDatang = LocalTime.parse(waktu, FORMAT_WAKTU);
    }

    public void setWaktuPulang(String waktu) {
        this.waktuPulang = LocalTime.parse(waktu, FORMAT_WAKTU);
    }

    public void setWaktuKerja(String datang, String pulang) {
        setWaktuDatang(datang);
        setWaktuPulang(pulang);
    }

    public String getNip() {
        return this.nip;
    }

    public String getNama() {
        return this.nama;
    }

    public int getGolongan() {
        return this.golongan;
    }

    public LocalTime getWaktuDatang() {
        return this.waktuDatang;
    }

    public LocalTime getWaktuPulang() {
        return this.waktuPulang;
    }

    public String getWaktuDatangTeks() {
        return (this.waktuDatang == null) ? "-" : this.waktuDatang.format(FORMAT_WAKTU);
    }

    public String getWaktuPulangTeks() {
        return (this.waktuPulang == null) ? "-" : this.waktuPulang.format(FORMAT_WAKTU);
    }

    public void inputPegawai(){
        Scanner scanner = new Scanner(System.in);
        System.out.print("Masukkan NIP: ");
        this.nip = scanner.nextLine();
        System.out.print("Masukkan Nama Pegawai: ");
        this.nama = scanner.nextLine();
        System.out.print("Masukkan Golongan (1/2/3/4): ");
        this.golongan = scanner.nextInt();
        scanner.nextLine();
        inputWaktu(scanner);
    }

    public void inputWaktu(Scanner scanner){
        this.waktuDatang = bacaWaktu(scanner, "Masukkan Waktu Datang (HH:mm): ");
        this.waktuPulang = bacaWaktu(scanner, "Masukkan Waktu Pulang (HH:mm): ");
    }

    private LocalTime bacaWaktu(Scanner scanner, String pesan){
        while (true) {
            System.out.print(pesan);
            String teks = scanner.nextLine().trim();
            try {
                return LocalTime.parse(teks, FORMAT_WAKTU);
            } catch (DateTimeParseException e) {
                System.out.println("Format waktu tidak valid. Contoh: 08:00 atau 17:30");
            }
        }
    }

    public Duration hitungLamaKerjaReturn() {
        if (this.waktuDatang == null || this.waktuPulang == null) {
            return Duration.ZERO;
        }
        Duration lama = Duration.between(this.waktuDatang, this.waktuPulang);
        if (lama.isNegative()) {
            lama = lama.plusHours(24);
        }
        return lama;
    }

    public String getLamaKerjaTeks() {
        Duration lama = hitungLamaKerjaReturn();
        return lama.toHours() + " jam " + lama.toMinutesPart() + " menit";
    }

    public void hitungLamaKerjaVoid() {
        if (this.waktuDatang == null || this.waktuPulang == null) {
            System.out.println(" Lama Kerja = - (waktu belum diisi)");
            return;
        }
        Duration lama = Duration.between(this.waktuDatang, this.waktuPulang);
        if (lama.isNegative()) {
            lama = lama.plusHours(24);
        }
        System.out.printf(" Lama Kerja = %d jam %d menit\n", lama.toHours(), lama.toMinutesPart());
    }

    public double cariGapok() {
        switch (this.golongan) {
            case 1: return 1500000;
            case 2: return 2000000;
            case 3: return 3000000;
            case 4: return 5000000;
            default: return 0;
        }
    }

    public double cariTunjangan() {
        switch (this.golongan) {
            case 1: return cariGapok() * 0.10;
            case 2: return cariGapok() * 0.12;
            case 3: return cariGapok() * 0.12;
            case 4: return cariGapok() * 0.15;
            default: return 0;
        }
    }

    public double cariPotongan() {
        switch (this.golongan) {
            case 1: return cariGapok() * 0.01;
            case 2: return cariGapok() * 0.02;
            case 3: return cariGapok() * 0.02;
            case 4: return cariGapok() * 0.04;
            default: return 0;
        }
    }

    public double cariGajiTotal() {
        return cariGapok() + cariTunjangan() - cariPotongan();
    }

    public double hitungGajiReturn() {
        return cariGapok() + cariTunjangan() - cariPotongan();
    }

    public void hitungGajiVoid() {
        double gapok = cariGapok();
        double tunjangan = cariTunjangan();
        double potongan = cariPotongan();
        double total = gapok + tunjangan - potongan;
        System.out.printf(" Gapok      = Rp %,.0f\n", gapok);
        System.out.printf(" Tunjangan  = Rp %,.0f\n", tunjangan);
        System.out.printf(" Potongan   = Rp %,.0f\n", potongan);
        System.out.printf(" Total Gaji = Rp %,.0f\n", total);
    }

    public void cetakWaktuKerja() {
        String datang = (this.waktuDatang == null) ? "-" : this.waktuDatang.format(FORMAT_WAKTU);
        String pulang = (this.waktuPulang == null) ? "-" : this.waktuPulang.format(FORMAT_WAKTU);
        String lama = (this.waktuDatang == null || this.waktuPulang == null) ? "-" : getLamaKerjaTeks();

        System.out.printf(" Waktu Datang (Clock In)  = %s\n", datang);
        System.out.printf(" Waktu Pulang (Clock Out) = %s\n", pulang);
        System.out.printf(" Lama Kerja               = %s\n", lama);
    }

    public void cetakPegawai() {
        System.out.println("-".repeat(112));
        System.out.printf("| %-15s | %-15s | %-3s | %-13s | %-13s | %-13s | %-15s |\n", 
                "NIP", "Nama", "Gol", "Gapok", "Tunjangan", "Potongan", "Total Gaji");
        System.out.println("-".repeat(112));
        System.out.printf("| %-15s | %-15s | %-3d | Rp %,10.0f | Rp %,10.0f | Rp %,10.0f | Rp %,12.0f |\n", 
                this.nip, this.nama, this.golongan, cariGapok(), cariTunjangan(), cariPotongan(), cariGajiTotal());
        System.out.println("-".repeat(112));
        cetakWaktuKerja();
    }
}

public class Nomor2_063_078_090{
    public static void main(String[] args){
        Pegawai obj1 = new Pegawai();
        Pegawai obj2 = new Pegawai("140810250090", "Razan", 3);
        Pegawai obj3 = new Pegawai();

        obj2.setWaktuKerja("08:00", "17:30");

        Pegawai[] daftar = {obj1, obj2, obj3};

        Scanner scanner = new Scanner(System.in);
        int pilihan;

        do{
            System.out.println("\n===== MENU GajiPegawai =====");
            System.out.println("1. Hitung Gaji & Lama Kerja (void)");
            System.out.println("2. Hitung Gaji & Lama Kerja (return)");
            System.out.println("3. Keluar");
            System.out.print("Pilih menu: ");
            pilihan = scanner.nextInt();
            scanner.nextLine();

            switch (pilihan){
                case 1:
                    System.out.println("Hitung (void)");
                    HitungVoid(scanner, daftar);
                    break;
                case 2:
                    System.out.println("Hitung (return)");
                    HitungReturn(scanner, daftar);
                    break;
                case 3:
                    System.out.println("Keluar dari program.");
                    break;
                default:
                    System.out.println("Pilihan tidak valid.");
            }
        }while(pilihan != 3);
    }


    static Pegawai pilihObjek(Scanner scanner, Pegawai[] daftar) {
        System.out.println("1. Obj. 1 (data diset lewat setter)");
        System.out.println("2. Obj. 2 (data diset lewat constructor)");
        System.out.println("3. Obj. 3 (data diinput dari keyboard)");
        System.out.print("Pilih objek (1-3): ");
        int no = scanner.nextInt();
        scanner.nextLine();

        switch (no) {
            case 1:
                daftar[0].setPegawai("140810250078", "Irsyad", 1);
                daftar[0].setWaktuKerja("07:45", "16:15");
                break;
            case 2:
                break;
            case 3:
                daftar[2].inputPegawai();
                break;
            default:
                System.out.println("Objek tidak valid.");
                return null;
        }
        return daftar[no - 1];
    }

    static void HitungVoid(Scanner scanner, Pegawai[] daftar) {
        Pegawai p = pilihObjek(scanner, daftar);
        if (p == null) return;
        System.out.printf(" NIP        = %s\n", p.getNip());
        System.out.printf(" Nama       = %s\n", p.getNama());
        System.out.printf(" Golongan   = %d\n", p.getGolongan());
        System.out.printf(" Clock In   = %s\n", p.getWaktuDatangTeks());
        System.out.printf(" Clock Out  = %s\n", p.getWaktuPulangTeks());
        p.hitungGajiVoid();
        p.hitungLamaKerjaVoid();
    }

    static void HitungReturn(Scanner scanner, Pegawai[] daftar) {
        Pegawai p = pilihObjek(scanner, daftar);
        if (p == null) return;
        double gapok = p.cariGapok();
        double tunjangan = p.cariTunjangan();
        double potongan = p.cariPotongan();
        double total = p.hitungGajiReturn();
        Duration lama = p.hitungLamaKerjaReturn();
        System.out.printf(" NIP        = %s\n", p.getNip());
        System.out.printf(" Nama       = %s\n", p.getNama());
        System.out.printf(" Golongan   = %d\n", p.getGolongan());
        System.out.printf(" Clock In   = %s\n", p.getWaktuDatangTeks());
        System.out.printf(" Clock Out  = %s\n", p.getWaktuPulangTeks());
        System.out.printf(" Gapok      = Rp %,.0f\n", gapok);
        System.out.printf(" Tunjangan  = Rp %,.0f\n", tunjangan);
        System.out.printf(" Potongan   = Rp %,.0f\n", potongan);
        System.out.printf(" Total Gaji = Rp %,.0f\n", total);
        System.out.printf(" Lama Kerja = %d jam %d menit\n", lama.toHours(), lama.toMinutesPart());
    }
}

File: Soal2/Nomor2_063_078_090.py
# Nama Program  :
# Nama Kelompok : Tim PeBO
# Nama Anggota  : Razan Ibrahim Nabil, Muhammah Irsyad Azzarul Haq, Djeremy Rieldy Marchiano Panjaitan
# NPM  Anggota  : 140810250090, 250078, 250063
# Tanggal Buat  : 24/09/2026
# Deskripsi     :

from datetime import date, datetime, timedelta

def parse_waktu(teks):
    """Pengganti LocalTime.parse(..., "HH:mm"). Mengembalikan datetime.time,
    atau melempar ValueError jika formatnya salah (seperti exception di Java)."""
    if len(teks) != 5:
        raise ValueError("Format waktu tidak valid: " + teks)
    return datetime.strptime(teks, "%H:%M").time()


def format_rupiah(nilai):
    """Pengganti "%,.0f" di Java: 3360000 -> "3.360.000" """
    return f"{nilai:,.0f}".replace(",", ".")


def jam_menit(lama):
    """Mengubah timedelta menjadi (jam, menit). Pengganti toHours() dan toMinutesPart()."""
    total_menit = int(lama.total_seconds() // 60)
    return total_menit // 60, total_menit % 60


def baca_int(pesan):
    """Pengganti scanner.nextInt(): diulang sampai yang diketik benar-benar angka."""
    while True:
        try:
            return int(input(pesan))
        except ValueError:
            print("Input harus berupa angka.")


def baca_waktu(pesan):
    """Membaca waktu dari keyboard, diulang sampai formatnya valid."""
    while True:
        teks = input(pesan).strip()
        try:
            return parse_waktu(teks)
        except ValueError:
            print("Format waktu tidak valid. Contoh: 08:00 atau 17:30")


class Pegawai:
    GAPOK = {1: 1500000, 2: 2000000, 3: 3000000, 4: 5000000}
    PERSEN_TUNJANGAN = {1: 0.10, 2: 0.12, 3: 0.12, 4: 0.15}
    PERSEN_POTONGAN = {1: 0.01, 2: 0.02, 3: 0.02, 4: 0.04}

    def __init__(self, nip="", nama="", golongan=0):
        self._nip = nip
        self._nama = nama
        self._golongan = golongan
        self._waktu_datang = None 
        self._waktu_pulang = None

    def set_pegawai(self, nip, nama, golongan):
        self._nip = nip
        self._nama = nama
        self._golongan = golongan

    def set_nip(self, nip):
        self._nip = nip

    def set_nama(self, nama):
        self._nama = nama

    def set_golongan(self, golongan):
        self._golongan = golongan

    def set_waktu_datang(self, waktu):
        self._waktu_datang = parse_waktu(waktu)

    def set_waktu_pulang(self, waktu):
        self._waktu_pulang = parse_waktu(waktu)

    def set_waktu_kerja(self, datang, pulang):
        self.set_waktu_datang(datang)
        self.set_waktu_pulang(pulang)

    # ---------- Getter ----------
    def get_nip(self):
        return self._nip

    def get_nama(self):
        return self._nama

    def get_golongan(self):
        return self._golongan

    def get_waktu_datang(self):
        return self._waktu_datang

    def get_waktu_pulang(self):
        return self._waktu_pulang

    def get_waktu_datang_teks(self):
        return "-" if self._waktu_datang is None else self._waktu_datang.strftime("%H:%M")

    def get_waktu_pulang_teks(self):
        return "-" if self._waktu_pulang is None else self._waktu_pulang.strftime("%H:%M")

    # ---------- Input ----------
    def input_waktu(self):
        self._waktu_datang = baca_waktu("Masukkan Waktu Datang (HH:mm): ")
        self._waktu_pulang = baca_waktu("Masukkan Waktu Pulang (HH:mm): ")

    def input_pegawai(self):
        self._nip = input("Masukkan NIP: ")
        self._nama = input("Masukkan Nama Pegawai: ")
        self._golongan = baca_int("Masukkan Golongan (1/2/3/4): ")
        self.input_waktu()

    def hitung_lama_kerja_return(self):
        if self._waktu_datang is None or self._waktu_pulang is None:
            return timedelta(0)
        hari_ini = date.today()
        lama = (datetime.combine(hari_ini, self._waktu_pulang)
                - datetime.combine(hari_ini, self._waktu_datang))
        if lama < timedelta(0):
            lama += timedelta(hours=24)
        return lama

    def get_lama_kerja_teks(self):
        jam, menit = jam_menit(self.hitung_lama_kerja_return())
        return f"{jam} jam {menit} menit"

    def hitung_lama_kerja_void(self):
        if self._waktu_datang is None or self._waktu_pulang is None:
            print(" Lama Kerja = - (waktu belum diisi)")
            return
        hari_ini = date.today()
        lama = (datetime.combine(hari_ini, self._waktu_pulang)
                - datetime.combine(hari_ini, self._waktu_datang))
        if lama < timedelta(0):
            lama += timedelta(hours=24)
        jam, menit = jam_menit(lama)
        print(f" Lama Kerja = {jam} jam {menit} menit")

    def cari_gapok(self):
        return self.GAPOK.get(self._golongan, 0)

    def cari_tunjangan(self):
        return self.cari_gapok() * self.PERSEN_TUNJANGAN.get(self._golongan, 0)

    def cari_potongan(self):
        return self.cari_gapok() * self.PERSEN_POTONGAN.get(self._golongan, 0)

    def cari_gaji_total(self):
        return self.cari_gapok() + self.cari_tunjangan() - self.cari_potongan()

    def hitung_gaji_return(self):
        return self.cari_gapok() + self.cari_tunjangan() - self.cari_potongan()

    def hitung_gaji_void(self):
        gapok = self.cari_gapok()
        tunjangan = self.cari_tunjangan()
        potongan = self.cari_potongan()
        total = gapok + tunjangan - potongan
        print(f" Gapok      = Rp {format_rupiah(gapok)}")
        print(f" Tunjangan  = Rp {format_rupiah(tunjangan)}")
        print(f" Potongan   = Rp {format_rupiah(potongan)}")
        print(f" Total Gaji = Rp {format_rupiah(total)}")

    def cetak_waktu_kerja(self):
        if self._waktu_datang is None or self._waktu_pulang is None:
            lama = "-"
        else:
            lama = self.get_lama_kerja_teks()
        print(f" Waktu Datang (Clock In)  = {self.get_waktu_datang_teks()}")
        print(f" Waktu Pulang (Clock Out) = {self.get_waktu_pulang_teks()}")
        print(f" Lama Kerja               = {lama}")

    def cetak_pegawai(self):
        print("-" * 112)
        print(f"| {'NIP':<15} | {'Nama':<15} | {'Gol':<3} | {'Gapok':<13} | "
              f"{'Tunjangan':<13} | {'Potongan':<13} | {'Total Gaji':<15} |")
        print("-" * 112)
        print(f"| {self._nip:<15} | {self._nama:<15} | {self._golongan:<3} | "
              f"Rp {format_rupiah(self.cari_gapok()):>10} | "
              f"Rp {format_rupiah(self.cari_tunjangan()):>10} | "
              f"Rp {format_rupiah(self.cari_potongan()):>10} | "
              f"Rp {format_rupiah(self.cari_gaji_total()):>12} |")
        print("-" * 112)
        self.cetak_waktu_kerja()


def pilih_objek(daftar):
    """Memilih objek 1-3 sekaligus menyiapkan datanya.
    Mengembalikan objek yang dipilih, atau None jika pilihan tidak valid."""
    print("1. Obj. 1 (data diset lewat setter)")
    print("2. Obj. 2 (data diset lewat constructor)")
    print("3. Obj. 3 (data diinput dari keyboard)")
    no = baca_int("Pilih objek (1-3): ")

    if no == 1:
        daftar[0].set_pegawai("140810250078", "Irsyad", 1)
        daftar[0].set_waktu_kerja("07:45", "16:15")
    elif no == 2:
        pass 
    elif no == 3:
        daftar[2].input_pegawai()
    else:
        print("Objek tidak valid.")
        return None
    return daftar[no - 1]


def hitung_void(daftar):
    """Cara 1: method void, hasil dicetak di dalam method."""
    p = pilih_objek(daftar)
    if p is None:
        return
    print(f" NIP        = {p.get_nip()}")
    print(f" Nama       = {p.get_nama()}")
    print(f" Golongan   = {p.get_golongan()}")
    print(f" Clock In   = {p.get_waktu_datang_teks()}")
    print(f" Clock Out  = {p.get_waktu_pulang_teks()}")
    p.hitung_gaji_void()
    p.hitung_lama_kerja_void()


def hitung_return(daftar):
    """Cara 2: method return, hasil diterima lalu dicetak di sini."""
    p = pilih_objek(daftar)
    if p is None:
        return
    gapok = p.cari_gapok()
    tunjangan = p.cari_tunjangan()
    potongan = p.cari_potongan()
    total = p.hitung_gaji_return()
    jam, menit = jam_menit(p.hitung_lama_kerja_return())
    print(f" NIP        = {p.get_nip()}")
    print(f" Nama       = {p.get_nama()}")
    print(f" Golongan   = {p.get_golongan()}")
    print(f" Clock In   = {p.get_waktu_datang_teks()}")
    print(f" Clock Out  = {p.get_waktu_pulang_teks()}")
    print(f" Gapok      = Rp {format_rupiah(gapok)}")
    print(f" Tunjangan  = Rp {format_rupiah(tunjangan)}")
    print(f" Potongan   = Rp {format_rupiah(potongan)}")
    print(f" Total Gaji = Rp {format_rupiah(total)}")
    print(f" Lama Kerja = {jam} jam {menit} menit")


def main():
    daftar = [
        Pegawai(),
        Pegawai("140810250090", "Razan", 3),
        Pegawai(),
    ]

    daftar[1].set_waktu_kerja("08:00", "17:30")

    pilihan = 0
    while pilihan != 3: 
        print("\n===== MENU GajiPegawai =====")
        print("1. Hitung Gaji & Lama Kerja (void)")
        print("2. Hitung Gaji & Lama Kerja (return)")
        print("3. Keluar")
        pilihan = baca_int("Pilih menu: ")

        if pilihan == 1:
            print("Hitung (void)")
            hitung_void(daftar)
        elif pilihan == 2:
            print("Hitung (return)")
            hitung_return(daftar)
        elif pilihan == 3:
            print("Keluar dari program.")
        else:
            print("Pilihan tidak valid.")


if __name__ == "__main__":
    try:
        main()
    except (EOFError, KeyboardInterrupt): 
        print("\nProgram dihentikan.")